### PR TITLE
[TRANSFORM] use precompiled JavaScript for faster transformation

### PR DESCRIPTION
**Pull request by THKDev**
On an ARM machine transformation by JavaScript [takes ~15 seconds each time](https://community.openhab.org/t/js-transformation-takes-10-15-seconds-each-call/61566). To speed up this a little bit, each JavaScript file will be pre compiled and cached. This reduces the time of transformation to ~500ms. First call of each JavaScript still takes ~15 seconds. each further call needs then ~500ms.
Transformation directory will be observed for file changes and the cache will be refreshed.

This implementation is similar to [org.eclipse.smarthome.automation.module.script](https://github.com/eclipse/smarthome/tree/master/bundles/automation/org.eclipse.smarthome.automation.module.script)

### CODE REVIEW

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /**
- * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ * Copyright (c) 2014,2019 Contributors to the Eclipse Foundation
```
[the remark this answers is not in the document]
Comment (kaikreuzer) [reply extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java]: I guess we have to...

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java [hunk 2]
```diff
@@ -37,61 +32,56 @@
...
-        Logger logger = LoggerFactory.getLogger(JavaScriptTransformationService.class);
-        if (filename == null || source == null) {
-            throw new TransformationException("the given parameters 'filename' and 'source' must not be null");
-        }
-
-        logger.debug("about to transform '{}' by the Java Script '{}'", source, filename);
-
-        Reader reader;
+        Objects.requireNonNull(filename, "the given parameter 'filename' must not be null");
+        Objects.requireNonNull(source, "the given parameter 'source' must not be null");
```
[the remark this answers is not in the document]
Comment (THKDev) [reply extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java]: It won't give a warning 
Multiple markers at this line
	- Redundant null check: comparing '@NonNull String' against null
	- Redundant null check: comparing '@NonNull String' against null
I will revert this.

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * The {@link TransformationScriptWatcher} watches the transformation directory for files. If a deleted/modified file is
+ * detected, the script is passed to the {@link JavaScriptEngineManager}.
+ *
+ * @author thomask
+ *
+ */
+@Component(immediate = true)
```
[the remark this answers is not in the document]
Comment (THKDev) [reply extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java]: I don't know. I took it from [ScriptFileWatcher](https://github.com/eclipse/smarthome/blob/16d61a1cb06ea7943a33b6b1d6840d5674c71b29/bundles/automation/org.eclipse.smarthome.automation.module.script.rulesupport/src/main/java/org/eclipse/smarthome/automation/module/script/rulesupport/internal/loader/ScriptFileWatcher.java#L52).

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptEngineManager.java [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    private final Logger logger = LoggerFactory.getLogger(JavaScriptEngineManager.class);
+    private final ScriptEngineManager manager = new ScriptEngineManager();
+    /* keep memory foot print low. max 2 concurrent threads are estimated */
+    private final Map<String, CompiledScript> compiledScriptMap = new ConcurrentHashMap<>(4, 0.5f, 2);
+
+    /**
+     *
+     * @param filename
+     * @return
+     * @throws TransformationException
```
[the remark this answers is not in the document]
Comment (maggu2810) [reply extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptEngineManager.java]: * The space between `CompiledScript` and `}` should be removed
* A `TransformationException` is thrown if ?

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (maggu2810) [reply extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java]: I will come up with a PR soon

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java [hunk 5]
```diff
@@ -12,36 +12,42 @@
...
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 /**
  * The implementation of {@link TransformationService} which transforms the
  * input by Java Script.
  *
- * @author Pauli Anttila
+ * @author Pauli Anttila Initial contribution
+ * @author Thomas Kordelle pre compiled scripts
```
Comment (maggu2810) [inline extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/JavaScriptTransformationService.java]: Thank you, I assume we use a `-` between name and description in the other files. Can you check?

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (maggu2810) [reply extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java]: `immediate = true` has been (mis)used. So if there is no need (I don't see any here), remove it.

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+import org.eclipse.smarthome.core.service.AbstractWatchService;
+import org.eclipse.smarthome.core.transform.TransformationService;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * The {@link TransformationScriptWatcher} watches the transformation directory for files. If a deleted/modified file is
+ * detected, the script is passed to the {@link JavaScriptEngineManager}.
+ *
+ * @author Thomas Kordelle pre compiled scripts
```
Comment (maggu2810) [inline extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java]: Add a dash if needed (see above)

File: extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        logger.debug("New watch event {} for path {}.", kind, path);
+
+        if (kind == OVERFLOW) {
+            return;
+        }
+
+        final WatchEvent<Path> ev = (WatchEvent<Path>) event;
+        final Path filename = ev.context();
+
+        logger.debug("Reloading javascript file {}.", filename.toString());
```
Comment (maggu2810) [inline extensions/transform/org.eclipse.smarthome.transform.javascript/src/main/java/org/eclipse/smarthome/transform/javascript/internal/TransformationScriptWatcher.java]: Don't call `toString` yourself on parameterized slf4j logging.

**Comment by maggu2810** [comment]
@THKDev I realized you missed to add a `Signed-off-by` footer. Please have a look at the contribution guidelines.

**Comment by kaikreuzer** [comment]
@THKDev Please ping us once you have addressed the comments, I'd love to merge this!

**Comment by THKDev** [comment]
@kaikreuzer 
Last remarks edited.

**Comment by THKDev** [comment]
@maggu2810 
toString() removed.

**Comment by kaikreuzer** [comment]
@maggu2810 Are all your comments addressed?

**Comment by maggu2810** [comment]
The "ECA Validation Tool" tells me that `thkdev@protonmail.com` is an unknown user.

**Comment by kaikreuzer** [comment]
@THKDev Please see https://github.com/eclipse/smarthome/blob/master/CONTRIBUTING.md#legal-considerations how to sign the ECA - thanks!

**Comment by THKDev** [comment]
@kaikreuzer 
Done.

**Pull request merged**